ocaml/parsing/ast_mapper.ml

```diff
@@ -1059,7 +1062,13 @@ let default_mapper =
         attr_loc = this.location this a.attr_loc
       }
     );
-    attributes = (fun this l -> List.map (this.attribute this) l);
+    attributes = (fun this l ->
+      List.filter_map (fun attr ->
+        let m, _ = ME.maybe_of_attrs [attr] in
+        match m with
+        | Some m -> this.modes this m |> ME.attr_of
+        | None -> Some (this.attribute this attr)) l);
+
```

Comment:
Here we are detecting modes directly under attributes. It might be more principled if we detect it at places that they can actually occur. Such contexts can be found by search `Mode_expr.maybe_of_attrs` in `ast_iterator.ml`.

In addition, mode coercion expression (`local_ expr`) encodes modes not in attributes but in extension node. I created https://github.com/ocaml-flambda/flambda-backend/pull/2331 to make it easier to handle.

Reply:
This should be unblocked now that https://github.com/ocaml-flambda/flambda-backend/pull/2335 is merged.

Reply:
Great!

I still find it a bit tricky to implement `ast_mapper` correctly here.

I agree what you suggested is a better approach (the current implementation also makes mode attributes visible when someone overrides the `attributes` function which is not ideal).

If we take the suggested approach similar to `ast_iterator`:
- look for mode expressions at typ/pat/value_bindings sites
- call `sub.attributes` on non-mode attributes and `sub.modes` on mode expressions
- combine everything back together into an attribute list

Then there's a concern around ordering when we recombine the attributes. Do we maintain an invariant about whether mode attributes should always come first/last when parsing (suspect this is already the case but is it something we want to be explicit about)?